File: bin/newman.py
#!/home/kamei/local/bin/python3
# -*- coding: utf-8 -*-
# -*- encoding: utf-8 -*-

from numpy import *

'''
$LOAD_PATH.push('.')
require 'Parser'
dataset = parse("groupbow-bow_e")
require"pp"
linkmat=Array.new(dataset.size).map!{Array.new(dataset.size,0)}
dataset.each{|i,v|
  v.each{|j| linkmat[i.to_i][j[0].to_i]=j[1].to_i}
}
n=open("userbow_v").readlines.map{|x|x.chomp}.freeze
#linkmat = dataset[1].clone.freeze
for i in 0..linkmat.size - 1
  for j in 0..linkmat[i].size - 1
    if linkmat[i][j] != linkmat[j][i] then
      raise "Error: Linkmat is directional."
    end
  end
end
'''
alllines=open("user-group_e").read()

'''
cluster = Hash.new([].freeze)
for i in 0..(n.size - 1)
  cluster[i] = Array.new if cluster[i] == [].freeze
  cluster[i] << n[i]
end

m = ((linkmat.flatten.to_a.inject(0) {|t,a| t + a}) / 2).freeze

e = Array.new(cluster.size) {|v| v = Array.new(cluster.size, 0.0)}
for i in 0..(e.size - 1)
  for j in 0..(e[i].size - 1)
    e[i][j] = linkmat[i][j].to_f / m.to_f
  end
end

a = Array.new(e.size, 0.0)
for i in 0..(a.size - 1)
  for k in 0..(e[i].size - 1)
    a[i] += e[i][k]
  end
end

dq = Array.new(e.size) {|v| v = Array.new(e.size, 0.0)}
for i in 0..(dq.size - 1)
  for j in 0..(dq[i].size - 1)
    dq[i][j] = 2.0 * (e[i][j] - (a[i] * a[j]))
  end
end


maxi = -1; maxj = -1
dqmax = -100000000.0 
for i in 0..(dq.size - 1)
  for j in 0..(dq[i].size - 1)
    next if i == j
    if dq[i][j] > dqmax then
      maxi = i
      maxj = j
      dqmax = dq[i][j]
    end
  end
end


cluster[maxj] += cluster[maxi]
cluster.delete(maxi)

x = cluster.clone
cluster.clear
x.each_with_index {|c, i| cluster[i] = c[1].clone}
x.clear

$maxq = -10000000.0
$maxcluster = []

while cluster.size > 1 do

  e.each {|v| v.clear}.clear
  a.clear
  dq.each {|v| v.clear}.clear
  
  e = Array.new(cluster.size) {|v| v = Array.new(cluster.size, 0.0)}
  cluster.each_with_index {|u, ii|
    cluster.each_with_index {|v, jj|

      edge = 0
      u[1].each {|i|
        v[1].each {|j|

          #edge += linkmat[i][j]
          edge += linkmat[n.index(i)][n.index(j)]
        }
      }

      e[ii][jj] = edge.to_f / m.to_f
      e[ii][jj] /= 2.0 if ii == jj
    }
  }

  a = Array.new(e.size, 0.0)
  for i in 0..(a.size - 1)
    for k in 0..(e[i].size - 1)
      a[i] += e[i][k]
    end
  end

  q = 0.0
  for i in 0..(e.size - 1)
    q += e[i][i] - (a[i] ** 2)
  end
  if q > $maxq then
    $maxq = q
    $maxcluster = cluster.clone
  end

  dq = Array.new(e.size) {|v| v = Array.new(e.size, 0.0)}
  for i in 0..(dq.size - 1)
    for j in 0..(dq[i].size - 1)
      dq[i][j] = 2.0 * (e[i][j] - (a[i] * a[j]))
    end
  end

  dqmax = -100000000.0 
  for i in 0..(dq.size - 1)
    for j in 0..(dq[i].size - 1)
      next if i == j
      if dq[i][j] > dqmax then
        maxi = i
        maxj = j
        dqmax = dq[i][j]
      end
    end
  end
 
  cluster[maxj] += cluster[maxi]
  cluster.delete(maxi)

  x = cluster.clone
  cluster.clear
  x.each_with_index {|c, i| cluster[i] = c[1].clone}
  x.clear
end

p $maxq
p $maxcluster

=begin
# Load graph-v 
@ts=open(ARGV[1]).readlines.map{|x|x.chomp}
$maxcluster.each{|k,v|
  @line=""
  v.each{|vv|@line+=@ts[vv]+":1 "}
  open("topic-bow_e","a"){|f|f.puts "#{k} #{@line}"}
}
=end
=begin
Dir.chdir('./out')
$dir, $file = File::split(ARGV[1])
#np=$file.gsub(/.node/, ".cluster")
#np=argv_node.gsub(/.node/, ".cluster")
np = $file + '.cluster'
#open('c_0.cluster', 'w') {|file|
open(np, 'w') {|file|

  $maxcluster.each {|k, v|

    file.puts 'Cluster ' + k.to_s
    #p 'Cluster ' + k.to_s
    v.each {|w|
      #printf("%s ", tags[w])
      file.puts "\t" + tags[w]
    }
    #printf("\n\n")
    file.puts ""
  }
}

#}
=end
'''

File: bin/newman.rb
#!/home/kamei/local/bin/ruby -Ku
# -*- encoding: utf-8 -*-
# -*- coding: utf-8 -*-

def parse path
  @h=Hash.new(0)
  open(path).readlines.map{|x|
    x.chomp.split(" ").reject{|v|v.empty?}
  }.each{|u|
    @h[u.shift]=u.map{|w|w.split(":")}
  }
  @h.clone
end

@ugh=parse("/home/kamei/workspace/dataset/head1/user-group_e")
@gbh=parse("/home/kamei/workspace/dataset/head1/groupname-bow_e")

$processed=0
@ugh.each{|user,group|
  p $processed+=1
 
  next if File.exists?("/home/kamei/workspace/dataset/cluster/#{@user}-bowcluster_e")
 
  @th=Hash.new(0)
  group.each{|g|
    # Access to Group-BOW of group"g[0]".
    @gbh[g[0]].each{|f|
      @from=f[0]
      @th[@from]=Hash.new(0) if @th[@from]==0
      
      @gbh[g[0]].each{|t|
        @to=t[0]
        next if @from==@to
        @th[@to]=Hash.new(0) if @th[@to]==0

        @th[@from].store(@to,1) unless @th[@from].has_key?(@to)
        @th[@to].store(@from,1) unless @th[@to].has_key?(@from)

        if @th[@from].has_key?(@to)
          @th[@from][@to]+=1
          @th[@to][@from]+=1
        end
      }
    }
  }

  @gv=@th.keys.sort.freeze
  next if @gv.empty?

  @lmat=Array.new(@gv.size){Hash.new(0)}
  @gv.each_with_index{|vv,i|
    @th[vv].each_with_index{|lw,j|
      @lmat[i].store(@gv.index(lw[0]),lw[1].to_f)
    }
  }

  @m=@lmat.map{|v|v.values.to_a}.flatten.inject(0){|t,a|t+a}.freeze

  @cluster=Array.new(@gv.size){|v|[v]}

  @qmax=-10000000
  @last=[]
  while @cluster.size > 1
    printf("%d\r", @cluster.size)

    @e_=Array.new(@cluster.size){Hash.new(0)}
    for i in 0..@cluster.size-1
      for j in 0..@cluster.size-1
        @linktotal=0
        @cluster[i].each{|vi|
          @cluster[j].each{|vj|
            @linktotal+=@lmat[vi][vj]
          }
        }
        @e_[i].store(j,@linktotal/@m) if @linktotal > 0
      end
    end

    @a_=Array.new(@cluster.size,0)
    for i in 0..@cluster.size-1
      @e_[i].each{|j,w| @a_[i]+=w}
    end

    @dqmax=[nil,nil,0]
    for i in 0..@cluster.size-1
      @e_[i].each{|j,w|
        @dq=2*(w-(@a_[i]*@a_[j]))
        @dqmax=[i,j,@dq] if @dqmax[2] < @dq
      }
    end

    begin
      @cluster[@dqmax[0]].each{|vi|
        @cluster[@dqmax[1]].each{|vj|
          @lmat[vi][vj]=@lmat[vj][vi]=0
        }
      }
      @cluster[@dqmax[1]]+=@cluster[@dqmax[0]]
      @cluster.delete_at(@dqmax[0])
      @q=0
      for i in 0..@cluster.size-1
        @e_[i].each{|j,w|
          @q+=@e_[i][i]-(@a_[i]**2)
        }
      end

      if @qmax < @q
        @qmax=@q
        @last=@cluster
      end
    rescue
      break
    end
  end

  open("/home/kamei/workspace/dataset/cluster/#{user}-bowcluster_e","w"){|f|
    @last.each{|cluster|
      @bow=Array.new(cluster.size)
      cluster.each_with_index{|bowid,i|
        @bow[i]=@gv[bowid]
      }
      f.puts @bow.join(" ")
    }
  }
}

